text/3691-trusted-publishing-cratesio.md

```diff
@@ -0,0 +1,384 @@
+
+Some services that use OIDC to authenticate use the _ID token_ itself to authorize actions.
+
+1. ID tokens will differ in their supported claims from publishing provider to publishing provider.
+    - This would complicate the authorization code on crates.io for actual API actions (e.g. publishing a crate through our existing APIs) as it would require logic for each publishing provider.
+2. Identity Providers are in control of the ID token and its contents. Unexpected changes to these contents could lead to authentication issues to crates.io.
+    - Exchanging the ID token for a crates.io token ensures we have consistency in the authorization logic of crates.io APIs.
+3. ID tokens would not benefit as much from Secret Scanning or revocation. They are JWTs with a varying format not as well-suited as the crates.io API token.
+    - By exchanging the ID token for a crates.io-specific access token, when the access token is leaked it can be quickly detected and revoked by the [existing Secret Scanning integrations](https://github.blog/changelog/2023-01-19-the-crate-io-registry-is-now-a-github-secret-scanning-integrator/).
+
+The OpenSSF Security Software Repositories Working Group [provides more detailed justification](https://repos.openssf.org/trusted-publishers-for-all-package-repositories#oidc-id-tokens-and-repository-controlled-tokens) to use the token exchange model were a valid ID token is exchanged for a package repository specific access token. 
+
+
+### What is the impact of not doing this?
+[impact-of-not-doing]: #impact-of-not-doing
+
+Foregoing the OIDC integration with trusted publishing providers would result in the status-quo. Authentication to crates.io from CI/CD systems would be facilitated via the current API token implementation.
+
+Crate authors who are more security-sensitive could using the expiring option on API tokens to further reduce potential impact of a leaked API token. However, these API tokens would need to be regularly regenerated in order to avoid interruption in release processes.
+
+These same API tokens are likely stored long term in a secrets manager like GitHub Actions Secrets or AWS Secrets Manager where a misconfiguration could lead to exposure of the secret.
+
+> [!NOTE] 
+> Other security challenges that would persist by not utilizing Trusted Publishing are detailed in the [Motivation](#motivation) section of this document.
+
+## Prior art
+[prior-art]: #prior-art
+
+Other package repositories have implemented similar support:
+
+- [PyPI's Trusted Publishing](https://docs.pypi.org/trusted-publishers/)
+- [RubyGem's Trusted Publishing](https://guides.rubygems.org/trusted-publishing/)
+- [Dart's Automated Publishing](https://dart.dev/tools/pub/automated-publishing)
+- [OSSF's Trusted Publishers For All Package Repositories Blog Post](https://repos.openssf.org/trusted-publishers-for-all-package-repositories)
+
+## Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+
```

Comment:
one minor unresolved question that I don't think is a blocker for the RFC though: how do we deal with user-related things in the publish endpoint? e.g. our publish rate limits (and overrides) are currently implemented per-user, but when publishing via OIDC there is no user. we also have a `version.published_by` column that refers to a `users.id` and other user-related columns that we might need to rethink when we implement this.

Reply:
Github actions at least has an actor field in their token pointing to the user initiating the workflow.

Reply:
interesting, I guess that could be sufficient for now. but I guess once we start to support a) different CI providers and/or b) different auth providers we will have to think about this anyway 😅